exporter/internal/queue/sized_channel.go

```diff
@@ -62,22 +62,19 @@ func (vcq *sizedChannel[T]) push(el T, size int64, callback func() error) error
 // pop removes the element from the queue and returns it.
 // The call blocks until there is an item available or the queue is stopped.
 // The function returns true when an item is consumed or false if the queue is stopped and emptied.
-// The callback is called before the element is removed from the queue. It must return the size of the element.
-func (vcq *sizedChannel[T]) pop(callback func(T) (size int64)) (T, bool) {
+func (vcq *sizedChannel[T]) pop() (T, bool) {
 	el, ok := <-vcq.ch
-	if !ok {
-		return el, false
-	}
-
-	size := callback(el)
+	return el, ok
+}
 
+// updateSize updates size. It should be called when an item is removed from the queue.
+func (vcq *sizedChannel[T]) updateSize(deltaSize int64) {
```

Comment:
I like how this PR simplifies the code a bit. However, we increase coupling here: clients are now responsible for managing the queue size after popping an item. Is there another way we can keep the queue size management inside `sizedChannel` and get the new pulling behavior we need?

Reply:
The awkwardness comes from that we only know the size of the item after reading is complete, and reading is slow so we want to parallelize.

In the new model, we want to send a cue to the next worker to standby once the current worker is able to acquire the item (but before actually starting the read). Putting that to pseudocode

```
_, item <- sizedChannel.ch // 1
sendCueToTheNextWorker() // 2
request = read() // 3
updateSize(size(request)) // 4
```

Now let's assume we are putting all above steps inside `sizedChannel`, then step 3 and 4 need to be callbacks passed from `queue` as a callback; step 2 need to be a callback passed from `queueBatcher` to `queue` then to `sizedChannel`. I find that too many layers intertwined together.

Let me know what you think - I am open to ideas